src/config.rs

```diff
@@ -0,0 +1,215 @@
+use std::collections::BTreeMap;
+use std::path::{Path, PathBuf};
+use std::fs::File;
+use std::io::Read;
+use toml::{self, Value};
+use serde::Deserialize;
+
+use errors::*;
+
+
+/// The overall configuration object for MDBook.
+#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
+#[serde(default)]
+pub struct Config {
+    /// Metadata about the book.
+    pub book: BookConfig,
+    /// Arbitrary information which renderers can use during the rendering
+    /// stage.
+    pub output: BTreeMap<String, Value>,
+    /// Information for use by preprocessors.
+    pub preprocess: BTreeMap<String, Value>,
+    /// Information for use by postprocessors.
+    pub postprocess: BTreeMap<String, Value>,
+}
+
+impl Config {
+    /// Load a `Config` from some string.
+    pub fn from_str(src: &str) -> Result<Config> {
+        toml::from_str(src).chain_err(|| Error::from("Invalid configuration file"))
+    }
+
+    /// Load the configuration file from disk.
+    pub fn from_disk<P: AsRef<Path>>(config_file: P) -> Result<Config> {
+        let mut buffer = String::new();
+        File::open(config_file)
+            .chain_err(|| "Unable to open the configuration file")?
+            .read_to_string(&mut buffer)
+            .chain_err(|| "Couldn't read the file")?;
+
+        Config::from_str(&buffer)
+    }
+
+    /// Convenience method for getting the html renderer's configuration.
+    ///
+    /// # Note
+    ///
+    /// This is for compatibility only. It will be removed completely once the
+    /// rendering and plugin system is established.
+    pub fn html_config(&self) -> Option<HtmlConfig> {
+        self.try_get_output("html").ok()
```

Comment:
This is not cached, so it will be deserialized a few times, but I think we do not need to optimize this.
And editing a variable in code is not as straight forward as with the `BookConfig`. Maybe you should add a line in the documentation on how to change a html config variable.

Reply:
Hmm... To be honest, I hadn't thought about letting people mutate the configuration after it has been loaded. 

I think I'll rewrite the config to internally hold the raw `Value`, with getters and setters for accessing things like an `output` table (e.g. `HtmlConfig`) by key. The `toml` crate has a [get_mut()] method which I could expose to people so they can get something by index (e.g. `output.html`) and possibly mutate it.

[get_mut()]: https://docs.rs/toml/0.4.5/toml/value/enum.Value.html#method.get_mut

Reply:
I just experimented with `Config` holding a raw `Value` and then providing getters and setters for the various components (book settings, outputs, etc) and it doesn't end up being overly ergonomic. The issue is that every time you want to access something from the `BookConfig` you need to deserialize it again, which requires dealing with errors appropriately. 

I think if I leave the internals `pub` that should be enough to let people update and inspect the config if they need to.

Reply:
I also tried a little but, but it did not end well, (see comment in line 57).
https://play.rust-lang.org/?gist=29a32500c428bfae20593421ec7bc4ab&version=stable

I would propose two other ways:
1. Do not use toml::Value at all and just use `HtmlConfig` as the only struct in `output` with direct deserialization. Add an API to get the original config string (raw file contents), so that the plugin can deserialize for itself.
2. Move the `HtmlConfig` struct into the `HtmlHandlebars`. Let the `HtmlHandlebars` deserialize from a `toml::Value` and store it's own deserialized config.

Reply:
In the long run I want to go for number 2 because that'll completely separate the handlebars renderer from the book. I think I marked the `HtmlConfig` getter as deprecated to indicate that.